### ISSUE TITLE
SSZipArchive logging errors because I deleted some of the unzipped files in delegate

### ISSUE THREAD

**Issue by alexhunsley** [ZipArchive/ZipArchive#405]
As part of my delegate callback in zipArchiveDidUnzipFileAtIndex:totalFiles:archivePath:unzippedFilePath:, I'm deleting some of the unzipped files after processing them.

SSZipArchive is not liking this:

```
[SSZipArchive] Error setting directory file modification date attribute: The file “48233662_9.txt” doesn’t exist.
[SSZipArchive] Set attributes failed for directory: /var/mobile/Containers/Data/Application/42BB3F6E-049A-4821-99A7-1BCCE356ABD3/Documents/48233662_10.txt.
```
Any way of stopping it doing this? I get a log entry for each file this happens to (but not a crash or anything).

Unless the official advice is to not delete any files until entire zip operation is finished?
But the delegate method seems like such a natural place to do it.

**Comment by Coeur** [comment]
I consider it as a bug: we should tell delegate only after all write operations are done, including attributes, otherwise the delegate may be unaware of actual date or file permission associated with the unzipped file.

**Comment by alexhunsley** [comment]
Very good point actually, yes, the file information should be complete when the delegate is called.

**Comment by Coeur** [comment]
Should be easy to fix. Pull request welcome if you want to get the "Contributor" tag :)

**Comment by alexhunsley** [comment]
Yeah I think I'll have a look at doing this :)

**Comment by alexhunsley** [comment]
I think the comments in the code about the dictionary `directoriesModificationDates` are a bit inaccurate: this dict seems to contain full paths of the files, but the comments describe the code as setting the modification times on folders, i.e:

```
    // The process of decompressing the .zip archive causes the modification times on the folders
    // to be set to the present time. So, when we are done, they need to be explicitly set.
    // set the modification date on all of the directories.
```

**Comment by alexhunsley** [comment]
I think the code around setting the folder names isn't doing what it's meant to, in any case. See the dict `directoriesModificationDates` - it contains paths of files, not folders, and the loop to set the date is setting the dates on those files, not the folders they are inside:
    
```
    // The process of decompressing the .zip archive causes the modification times on the folders
    // to be set to the present time. So, when we are done, they need to be explicitly set.
    // set the modification date on all of the directories.
    if (success && preserveAttributes) {
        NSError * err = nil;
        for (NSDictionary * d in directoriesModificationDates) {
            if (![[NSFileManager defaultManager] setAttributes:@{NSFileModificationDate: d[@"modDate"]} ofItemAtPath:d[@"path"] error:&err]) {
                NSLog(@"[SSZipArchive] Set attributes failed for directory: %@.", d[@"path"]);
            }
            if (err) {
                NSLog(@"[SSZipArchive] Error setting directory file modification date attribute: %@", err.localizedDescription);
            }
        }
    }

```

**Comment by Coeur** [comment]
I don't know. Note that there are two ways of storing a folder with a file in a zip archive:

Way 1:
* by storing entry "folderPath/"
* and storing entry "folderPath/filename"

Way 2:
* by just storing entry "folderPath/filename"

So sometimes the date of the folder is given explicitely (Way 1), and sometimes we need to refer to the date of the most recent file in the folder (Way 2).

**Comment by Coeur** [comment]
Postponing to 2.1.3.

**Comment by Coeur** [comment]
> I think the code around setting the folder names isn't doing what it's meant to, in any case.

You're right. Seems like an issue in the original implementation in c05ee25f7f5411ceb5ec1099c79071c672746afc
